pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/MetadataStoreFactoryImpl.java

```diff
@@ -45,19 +62,50 @@ public static MetadataStoreExtended createExtended(String metadataURL, MetadataS
     private static MetadataStore newInstance(String metadataURL, MetadataStoreConfig metadataStoreConfig,
                                              boolean enableSessionWatcher)
             throws MetadataStoreException {
+        MetadataStoreProvider provider = findProvider(metadataURL);
+        if (provider != null) {
+            return provider.create(metadataURL, metadataStoreConfig, enableSessionWatcher);
+        }
+        return new ZKMetadataStore(metadataURL, metadataStoreConfig, enableSessionWatcher);
+    }
+
+    static void loadProviders() {
+        String factoryClasses = System.getProperty(METADATASTORE_PROVIDERS_PROPERTY);
+        providers = new HashMap<>();
+        providers.put(MEMORY_SCHEME_IDENTIFIER, new DefaultMetadataStoreProvider(MEMORY_SCHEME_IDENTIFIER,
+                (url, config, enableWatchers) -> new LocalMemoryMetadataStore(url, config)));
+        providers.put(ROCKSDB_SCHEME_IDENTIFIER, new DefaultMetadataStoreProvider(ROCKSDB_SCHEME_IDENTIFIER,
+                (url, config, enableWatchers) -> RocksdbMetadataStore.get(url, config)));
+        providers.put(ETCD_SCHEME_IDENTIFIER, new DefaultMetadataStoreProvider(ETCD_SCHEME_IDENTIFIER,
+                EtcdMetadataStore::new));
+        providers.put(ZK_SCHEME_IDENTIFIER, new DefaultMetadataStoreProvider(ZK_SCHEME_IDENTIFIER,
+                (url, config, enableWatchers) ->
+                        new ZKMetadataStore(url.substring(ZK_SCHEME_IDENTIFIER.length()), config, enableWatchers)));
+
+        if (factoryClasses == null) {
+            return;
+        }
+        String[] classNames = factoryClasses.split(",");
```

Comment:
Let's be more user-friendly - Trim spaces before or after the separator. Guava has great utility for that.
Let's ignore empty values if placed by mistake.,